feat: Implement automatic host reassignment when host leaves

balaraj74 opened the pull request:
# {PR Name}
Closes #1
## Which issue(s) does this PR address?
Right now, if the host leaves or disconnects, everyone gets kicked out and the jam ends. That's pretty annoying when you're in the middle of a session! This fix makes sure the jam keeps going by automatically making someone else the host.

## Why do we need this PR?
On the server side:

We now keep track of who's hosting each room
When the host leaves, we pick the next person in line to be the new host
Everyone gets notified about who the new host is
On the client side:
## What logical changes are present in this PR?
When someone joins as host, we let the server know
If you get promoted to host, you'll see a popup telling you## Is there some additional work to be done later that is NOT covered in this PR?
give this
The new host can then start streaming audio to everyone else
## How did you test the changes in this PR?

I ran the server locally and made sure:
## Are there any breaking changes in this PR?
Everything starts up properly
The server correctly tracks who's the host
When the host disconnects, someone else gets promoted
The new host gets notified and can take over

Nope! Everything that worked before still works the same way.

Yeah, a few nice-to-haves for later:
## Is there some additional work to be done later that is NOT covered in this PR?
give this
Better UI showing who's currently hosting
Let hosts manually pass control to someone else
Add some automated tests to make sure this keeps working

balaraj74 committed:
feat: Implement automatic host reassignment when host leaves

- Track hostId in server room objects to identify current host
- Automatically promote first remaining participant when host disconnects
- Notify all clients with 'host-promoted' message including new host details
- Enable promoted joiners to acquire media and stream audio as new host
- Add role-based permissions so only host can start/end calls
- Maintain session continuity instead of closing when host leaves

Fixes #1

balaraj74 committed:
Merge pull request #1 from balaraj74/feature/host-reassignment

feat: Implement automatic host reassignment when host leaves

maayamohan commented:
@balaraj74 where are you? I wanna discuss this in person.

balaraj74 commented:
I am in sem hall 1  5th row

maayamohan commented:
okk coming

maayamohan commented:
im in your row, in an acm hoodie

balaraj74 commented:
where are you

maayamohan commented:
@balaraj74 im so sorry i went to sem hall 2 (accidentally instead of sem hall 1). Can you come here

balaraj74 commented:
ok i will

balaraj74 commented:
I am in sem hall 2

maayamohan commented:
> I am in sem hall 2

where?

maayamohan commented:
im in 5th row, wearing acm hoodie

balaraj74 commented:
Where are you

maayamohan commented:
are u in sem hall 2

balaraj74 commented:
No I came back

balaraj74 commented:
I search everywhere

balaraj74 commented:
Can I upload the screenshot for the review insted

balaraj74 commented:
<img width="1916" height="921" alt="Screenshot from 2025-10-18 05-12-21" src="https://github.com/user-attachments/assets/f388cba3-a78d-4064-8e5c-a897112b866d" />
<img width="1916" height="921" alt="Screenshot from 2025-10-18 05-12-36" src="https://github.com/user-attachments/assets/dcd2554b-5eb1-4ca8-93fe-08095acb9444" />
<img width="1916" height="921" alt="Screenshot from 2025-10-18 05-12-55" src="https://github.com/user-attachments/assets/23c71134-9938-4429-8ba5-27a4ae41c395" />
<img width="1916" height="921" alt="Screenshot from 2025-10-18 05-13-05" src="https://github.com/user-attachments/assets/7d8e044d-5b5f-4da2-bbf6-1072ffbc54e5" />

balaraj74 committed:
fix: Show proper UI controls when participant is promoted to host

- Display local audio wrapper and hide remote audio for new host
- Show start and end buttons, hide exit button
- Enable start button with updated text "Start Jam (You are Host)"
- Add proper event listener to acquire audio and stream to participants
- Fix button references at top of join.js file

This ensures promoted hosts can actually stream audio instead of just
receiving a notification without controls.

balaraj74 committed:
fix: Promote next participant when host clicks "End Jam"

Previously, host promotion only worked when host disconnected/closed tab.
Now when host clicks "End Jam" button:
- Server removes host from room
- Broadcasts client-left message
- Promotes first remaining participant to host
- Sends host-promoted notification to all clients

This ensures session continuity whether host disconnects or ends session properly.

balaraj74 committed:
fix: Add comprehensive logging and improve host promotion UI handling

- Added detailed console.log statements throughout host promotion flow
- Get fresh references to DOM elements when promoted to avoid stale references
- Use inline-block display for buttons to ensure visibility
- Added server-side logging to track promotion messages being sent
- Created TESTING_GUIDE.md with step-by-step test instructions
- Improved error handling and user feedback

This will help debug and verify that host promotion is working correctly.

balaraj74 committed:
feat: Implement host reassignment with automatic promotion and UI updates

- Server-side: Track host in active rooms, automatically promote first participant when host leaves (disconnect or end-call)
- Server broadcasts room-update with participants and hostId to all clients
- Client-side: Add helper functions for room state caching, UI toggles (enableHostUI/enableListenerUI)
- Client-side: Handle host-promoted and room-update messages to sync host state
- Client-side: Promoted participant gets host controls (Start/End Jam buttons) with proper label
- UI: Display participants list with crown (👑) icon for current host
- UI: Listener participants show Exit button, hosts show Start/End Jam buttons
- Enhanced logging for debugging room state transitions

balaraj74 committed:
Merge pull request #2 from balaraj74/feature/host-reassignment

feat: Implement host reassignment with automatic promotion
